### PR TITLE
Fix race condition on resgroup slot reuse

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
New test cases are not provided as installcheck-resgroup can already trigger this issue with high failrate on low end system, like the one used by the resgroup concourse pipeline.

File: src/backend/utils/resgroup/resgroup.c [hunk 1]
```diff
@@ -2738,15 +2816,116 @@ procWakeup(PGPROC *proc)
...
 	return slot->groupId != InvalidOid;
 }
 
 /*
+ * An allocated slot is idle if it has an invalid groupId.
+ *
+ * The slot must be allocated on QD, no such requirement on QE.
+ */
+static bool
+slotIsIdle(const ResGroupSlotData *slot)
```
Comment (xiong-gang) [inline src/backend/utils/resgroup/resgroup.c]: why need two function `slotIsInUse` and `slotIsIdle`

File: src/backend/utils/resgroup/resgroup.c [hunk 2]
```diff
@@ -1141,7 +1167,12 @@ selfDetachSlot(ResGroupData *group, ResGroupSlotData *slot)
 static void
 initSlot(ResGroupSlotData *slot, ResGroupCaps *caps, Oid groupId, int sessionId)
 {
-	Assert(!slotIsInUse(slot));
+	Assert(slotIsIdle(slot));
+	Assert(caps != NULL);
+	Assert(groupId != InvalidOid);
+	Assert(sessionId != InvalidSessionId);
+
+	/* TODO: check for LWLock */
```
Comment (xiong-gang) [inline src/backend/utils/resgroup/resgroup.c]: init `prev` and `next`?

File: src/backend/utils/resgroup/resgroup.c (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: With this PR the slot can be in three states:

1. freed: unallocated/freed: it's still in the free list and not being used by any segment;
2. idle: allocated but not attached to any segment: already got by some segment but not attached yet, nProcs == 0, groupId == InvalidOid, etc.;
3. in use: allocated and attached to some segment(s): already got by some segment and attached to at least one of them, nProcs > 0, groupId != InvalidOid, etc.;

On QD a slot has all the 3 states, on QE it has only 2 (idle and in use). In the past we only have two states (in use and not in use). This new state on QD is an attempt for debugging purpose, we want to detect for potential race condition issue whenever possible.

File: src/backend/utils/resgroup/resgroup.c (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/resgroup/resgroup.c]: In the 2nd commit we added an assertion on `prev` and `next`.

**Comment by pivotal-ning-yu** [comment]
Per offline discussion this got approved by @xiong-gang, we want to have this merged asap as it is a quick fix for the red resgroup pipeline, due to a race condition issue introduced recently.

**Pull request merged**